### PR TITLE
New/fql

### PR DESCRIPTION
**Pull request by Deepak-Vohra**
Added a page on Full Query Logging, a new feature.

### CODE REVIEW

File: doc/source/new/fqllogging.rst [hunk 1]
```diff
@@ -0,0 +1,2096 @@
...
+
+Cassandra 4.0 includes an implementation of a full query logging (FQL) that uses chronicle-queue to implement a rotating log of queries. Some of the features of FQL are:
+
+- Single thread asynchronously writes log entries to disk to reduce impact on query latency
+- Heap memory usage bounded by a weighted queue with configurable maximum weight sitting in front of logging thread
+- If the weighted queue is full producers can be blocked or samples can be dropped
+- Disk utilization is bounded by deleting old log segments once a configurable size is reached
+- The on disk serialization uses a flexible schema binary format (chronicle-wire) making it easy to skip unrecognized fields, add new ones, and omit old ones.
+- Can be enabled and configured via JMX, disabled, and reset (delete on disk data), logging path is configurable via both JMX and YAML
+- Introduce new ``fqltool`` in ``/bin`` that currently implements Dump which can dump in a readable format full query logs as well as follow active full query logs
```
Comment (yifan-c) [inline doc/source/new/fqllogging.rst]: The fql tool set now includes `Dump`, `Replay` and `Compare`.

File: doc/source/new/fqllogging.rst [hunk 2]
```diff
@@ -0,0 +1,2096 @@
...
+
+Cassandra 4.0 has a binary full query log based on Chronicle Queue that can be controlled using ``nodetool enablefullquerylog``, ``disablefullquerylog``, and ``resetfullquerylog``. The log contains all queries invoked, approximate time they were invoked, any parameters necessary to bind wildcard values, and all query options. A readable version of the log can be dumped or tailed using the new ``bin/fqltool`` utility. The full query log is designed to be safe to use in production and limits utilization of heap memory and disk space with limits you can specify when enabling the log.
+
+Objective
+^^^^^^^^^^ 
+Full Query Logging logs all requests to the CQL interface. The full query logs could be used for debugging, performance benchmarking,  testing and auditing CQL queries. The audit logs also include CQL requests but full query logging is dedicated to CQL requests only with features such as FQL replay and FQL compare that are not available in audit logging. Audit logging is for auditing the database activity and may be improved to add auditing of other database activity beside authorization and CQL in future versions.
```
Comment (yifan-c) [inline doc/source/new/fqllogging.rst]: > Audit logging is for auditing the database activity and may be improved to add auditing of other database activity beside authorization and CQL in future versions.

The introduction of `Audit logging` does not align with the topic of FQL introduction.

Comment (yifan-c) [reply doc/source/new/fqllogging.rst]: > `performance benchmarking,  testing`

Remove extra space.

File: doc/source/new/fqllogging.rst [hunk 3]
```diff
@@ -0,0 +1,2096 @@
...
+
+Objective
+^^^^^^^^^^ 
+Full Query Logging logs all requests to the CQL interface. The full query logs could be used for debugging, performance benchmarking,  testing and auditing CQL queries. The audit logs also include CQL requests but full query logging is dedicated to CQL requests only with features such as FQL replay and FQL compare that are not available in audit logging. Audit logging is for auditing the database activity and may be improved to add auditing of other database activity beside authorization and CQL in future versions. 
+
+Full Query Logger
+^^^^^^^^^^^^^^^^^^ 
+The Full Query Logger is a logger that logs entire query contents after the query finishes (or times out). Queries are logged in one of two modes: single query or batch of queries. The log for an invocation of a batch of queries includes the following attributes:
```
Comment (yifan-c) [inline doc/source/new/fqllogging.rst]: FQL only logs the queries that successfully complete. The other queries (e.g. timed out, failed) are not to be logged.

File: doc/source/new/fqllogging.rst [hunk 4]
```diff
@@ -0,0 +1,2096 @@
...
+::
+
+ type - The type of the batch
+ queries - CQL text of the queries
+ values - Values to bind to as parameters for the queries
+ queryOptions - Options associated with the query invocation
+ queryState - Timestamp state associated with the query invocation
+ batchTimeMillis - Approximate time in milliseconds since the epoch since the batch was invoked
+
+Bin log is a  quick and dirty binary log that is kind of a NIH version of binary logging with a traditional logging framework. It's goal is good enough performance, predictable footprint, simplicity in terms of implementation and configuration and most importantly minimal impact on producers of log records. Performance safety is accomplished by feeding items to the binary log using a weighted queue and dropping records if the binary log falls sufficiently far behind. Simplicity and good enough performance is achieved by using a single log writing thread as well as Chronicle Queue to handle writing the log, making it available for readers, as well as log rolling.
```
Comment (yifan-c) [inline doc/source/new/fqllogging.rst]: `NIH` most likely stands for `not invented here`. However, it does not help to understand, although the abbreviation exist in the source code comment. I would suggest to just drop the clause.

`Bin log` was only used once in the entire document. How is it associate with FQL? 

`It's goal is` should be `Its goal is` or just `The goal is`

Probably have the following

> Full query logging is backed up by `BinLog`. `BinLog` is a quick and dirty binary log. Its goal is...

File: doc/source/new/fqllogging.rst (hunk 1 above)
Comment (Deepak-Vohra) [reply doc/source/new/fqllogging.rst]: Mentioned in next paragraph.

"The audit logs also include CQL requests but full query logging is dedicated to CQL requests only with features such as FQL Replay and FQL Compare that are not available in audit logging."

File: doc/source/new/fqllogging.rst [hunk 5]
```diff
@@ -0,0 +1,2096 @@
...
+
+Weighted queue is a wrapper around any blocking queue that turns it into a blocking weighted queue. The queue will weigh each element being added and removed. Adding to the queue is blocked if adding would violate the weight bound. If an element weighs in at larger than the capacity of the queue then exactly one such element will be allowed into the queue at a time. If the weight of an object changes after it is added it could create issues. Checking weight should be cheap so memorize expensive to compute weights. If weight throws that can also result in leaked permits so it's always a good idea to memorize weight so it doesn't throw. In the interests of not writing unit tests for methods no one uses there is a lot of ``UnsupportedOperationException``. If you need them then add them and add proper unit tests to ``WeightedQueueTest``. "Good" tests. 100% coverage including exception paths and resource leaks.
+
+The log for single CQL query includes the following attributes:
```
Comment (yifan-c) [inline doc/source/new/fqllogging.rst]: This should be sitting together with the one regarding batch query attributes above.

File: doc/source/new/fqllogging.rst [hunk 6]
```diff
@@ -0,0 +1,2096 @@
...
+
+Weighted queue is a wrapper around any blocking queue that turns it into a blocking weighted queue. The queue will weigh each element being added and removed. Adding to the queue is blocked if adding would violate the weight bound. If an element weighs in at larger than the capacity of the queue then exactly one such element will be allowed into the queue at a time. If the weight of an object changes after it is added it could create issues. Checking weight should be cheap so memorize expensive to compute weights. If weight throws that can also result in leaked permits so it's always a good idea to memorize weight so it doesn't throw. In the interests of not writing unit tests for methods no one uses there is a lot of ``UnsupportedOperationException``. If you need them then add them and add proper unit tests to ``WeightedQueueTest``. "Good" tests. 100% coverage including exception paths and resource leaks.
```
Comment (yifan-c) [inline doc/source/new/fqllogging.rst]: The introduction of the weighted queue is too detailed (especially the exceptions, the unit tests, etc.) that I feel only Cassandra developer will be interested in. Could we have a high-level introduction about the weighted queue, and move the details into a quoted note section for developers?

File: doc/source/new/fqllogging.rst [hunk 7]
```diff
@@ -0,0 +1,2096 @@
...
+
+The log for single CQL query includes the following attributes:
+
+::
+
+ query CQL query text
+ queryOptions Options associated with the query invocation
+ queryState Timestamp state associated with the query invocation
+ queryTimeMillis Approximate time in milliseconds since the epoch since the batch was invoked
```
Comment (yifan-c) [inline doc/source/new/fqllogging.rst]: The format is not consistent with batch query attributes. Please unify
 
`FieldName - Description`

File: doc/source/new/fqllogging.rst [hunk 8]
```diff
@@ -0,0 +1,2096 @@
...
+
+ full_query_logging_options:
+    # log_dir:
+
+Create the FQL directory if  it does not exist and set its permissions.
+
+::
+
+ sudo mkdir -p /tmp/cassandrafullquerylog 
+ sudo chmod -R 777 /tmp/cassandrafullquerylog
```
Comment (yifan-c) [inline doc/source/new/fqllogging.rst:130]: Why `777`? No files are expected to be executed.

File: doc/source/new/fqllogging.rst [hunk 9]
```diff
@@ -0,0 +1,2096 @@
...
+      WARN  [main] 2019-08-02 03:07:53,635 Slf4jExceptionHandler.java:42 - Using Pauser.sleepy() as not enough processors, have 2, needs 8+
+      Type: single-query
+      Query start time: 1564708322030
+      Protocol version: 4
+      Generated timestamp:-9223372036854775808
+      Generated nowInSeconds:1564708322
+      Query: SELECT * FROM system.peers
+      Values: 
+      
+      Type: single-query
```
Comment (yifan-c) [inline doc/source/new/fqllogging.rst:319]: Can we truncate the entries? It makes the page unnecessarily long.

File: doc/source/new/fqllogging.rst (hunk 1 above)
Comment (yifan-c) [reply doc/source/new/fqllogging.rst]: If the `dump` is mentioned as an item in the list, the other 2 should be added too.

File: doc/source/new/fqllogging.rst [hunk 10]
```diff
@@ -0,0 +1,705 @@
...
+
+
+The FQL tracks information about store files:
+
+- Store files as they are added and their storage impact. Delete them if over storage limit.
+- The files in the chronicle queue that have already rolled
+- The number of bytes in store files that have already rolled
+
+FQL logger sequence is as follows;
```
Comment (yifan-c) [inline doc/source/new/fqllogging.rst]: `follows;` -> `follows:`

**Review by yifan-c — COMMENTED** [review]
Appreciate the efforts on making those improvements! 
If possible, please further reduce the dumped content from the cq4 file.

**Comment by Deepak-Vohra** [comment]
- Removed a few more duplicate entries.
-  As the page is on FQL, best to list what all FQL provides even if takes slightly more page. 
- Added a comment that the log is truncated. 
- Made all references to "as follows" consistent by adding : suffix.

File: doc/source/new/fqllogging.rst [hunk 11]
```diff
@@ -0,0 +1,689 @@
...
+::
+
+ type - The type of the batch
+ queries - CQL text of the queries
+ values - Values to bind to as parameters for the queries
+ queryOptions - Options associated with the query invocation
+ queryState - Timestamp state associated with the query invocation
+ batchTimeMillis - Approximate time in milliseconds since the epoch since the batch was invoked
+ 
+ The log for single CQL query includes the following attributes:
```
Comment (yifan-c) [inline doc/source/new/fqllogging.rst]: Please remove the heading spaces so this line is not included in the code block

**Comment by dineshjoshi** [comment]
This has been committed.

**Pull request closed without merging**